### PR TITLE
Introduce skip_cache_func

### LINKED ISSUE #649: Block writing cache in `cached` decorator if custom condition is not satisfied
I m using the  `aiocache.cached` decorator with wraps the certain processing function. The setup is as following:

```python
import aiocache


aiocache.caches.add(
    "cache-alias",
    {
        "cache": "aiocache.RedisCache",
        {
            <cache-config-data>
        }
    },
)


@aiocache.cached(
    alias="cache-alias",
    key_builder=some_key_builder,
    ttl=ttl,
)
async def processing(...) -> int:
    ...
```

The problem is -- can i pass any argument to cached decorator in order to avoid cache writing for `processing(...)` return?
It seems like the only way to perform it is to throw exception inside the `processing(...)` function if the result is not appropriate one. I m looking for more generic way to do the same stuff.

For example, pass some `Callable` object to `cached.__call__()` returning the bool var meaning caching the results. Something similar to:

```python
def checker(value: Any) -> bool:
    ...

@aiocache.cached(
    alias="cache-alias",
    key_builder=some_key_builder,
    value_checker=checker,
    ttl=ttl,
)
async def processing(...) -> int:
    ...
```

### PR DESCRIPTION
**Pull request by VitalyPetrov**
- add an optional `value_checker` param to cache decorator classes `cached`, `cached_stampede` and `multi_cached`
- add IDE-specific folders to `gitignore`
- add black formatter to `requirements-dev.txt

## Are there changes in behavior for the user?

Did not test in real production code

## Related issue number

Fixes #649

## Checklist

- [+] I think the code is well written
- [-] Unit tests for the changes exist
- [+] Documentation reflects the changes
- [-] Add a new news fragment into the `CHANGES` folder

### CODE REVIEW

**Comment by VitalyPetrov** [comment]
Unit tests for that functionality is not done yet.
As for notes in `CHANGES` the PR seems to be too small to be released separately IMO. But the choice Is yours (maintainers).

File: aiocache/decorators.py [hunk 1]
```diff
@@ -58,6 +62,7 @@ def __init__(
         ttl=SENTINEL,
         namespace=None,
         key_builder=None,
+        value_checker=None,
```
Comment (Dreamsorcerer) [inline aiocache/decorators.py:65]:
```suggestion
        skip_cache_func=lambda r: False,
```

How about we just put a default function and not need to test for `None`?

File: aiocache/decorators.py [hunk 2]
```diff
@@ -111,6 +117,13 @@ async def decorator(
 
         result = await f(*args, **kwargs)
 
+        if self.value_checker:
+            if not isinstance(self.value_checker(result), bool):
+                raise TypeError('value_checker callable should return bool')
```
Comment (Dreamsorcerer) [inline aiocache/decorators.py:122]: I'd be inclined to leave this to type checkers (e.g. mypy) and not care about runtime checks here.

File: requirements-dev.txt [hunk 3]
```diff
@@ -1,5 +1,6 @@
 -r requirements.txt
 
+black==22.12.0
```
Comment (Dreamsorcerer) [inline requirements-dev.txt:3]: Not relevant to the PR.

```suggestion
```

File: aiocache/decorators.py [hunk 4]
```diff
@@ -268,6 +292,11 @@ class multi_cached:
         ``keys_from_attr``, the decorated callable, and the positional and keyword arguments
         that were passed to the decorated callable. This behavior is necessarily different than
         ``BaseCache.build_key()`` and the call signature differs from ``cached.key_builder``.
+    :param value_checker: dict of callables that allows to control which keys should not
+        be cached based on corresponding value check. The dict keys are set of keys to
+        be checked while the dict value is callable which takes corresponding result value
+        (result[key])  and returns boolean indicating whether cache that (key, value)
+        pair or not.
```
Comment (Dreamsorcerer) [inline aiocache/decorators.py:299]: Hmm, I'm not sure about this one, it seems to be less flexible than the others.
This requires knowing which keys should (or could) be cached. e.g. You couldn't skip keys that ended with `_nocache`, for example, unless you knew all possible keys beforehand.

I can think of 2 flexible options:
```
skip_keys = skip_cache_func(result.copy())
results_to_cache = {k: v for k, v in result.items() if k not in skip_keys}

# OR

results_to_cache = dict(it for it in result.items() if not skip_cache_func(*it))
```

**Review by Dreamsorcerer — COMMENTED** [review]
Thanks.

A couple of suggestions there, and we'll also need some tests for the new behaviour (maybe in tests/acceptance/tests_decorator.py so it gets tested against the different backends).

**Comment by Dreamsorcerer** [comment]
Ignore the news fragment. I'm not sure where that checklist comes from, trying to see if I can edit it...

File: aiocache/decorators.py (hunk 1 above)
Comment (VitalyPetrov) [reply aiocache/decorators.py:65]: Its an antipattern
https://www.flake8rules.com/rules/E731.html

Comment (Dreamsorcerer) [reply aiocache/decorators.py:65]: The function returns True, it's never going to appear in a traceback..

File: .gitignore [hunk 5]
```diff
@@ -86,6 +86,10 @@ ENV/
 # Spyder project settings
 .spyderproject
 
+# IDE project files
+.idea/
+.vscode/
+
```
[the remark this answers is not in the document]
Comment (VitalyPetrov) [reply .gitignore:92]: Yes, that fine, but allows from other dev to make the same settings on his machine. Local `.gitignore` is more flexible  here as you can be sure that there will be no extra files in your repo.

File: aiocache/decorators.py (hunk 1 above)
Comment (Dreamsorcerer) [reply aiocache/decorators.py:65]: Also, the antipattern is assigning to a variable. lambdas exist to be passed directly into a function.
`foo(lambda: ...)` is allowed
`f = lambda: ...` is not

File: .gitignore (hunk 5 above)
[the remark this answers is not in the document]
Comment (Dreamsorcerer) [reply .gitignore:92]: Expecting every project in the world to add ignore patterns for every IDE and tool in the world is not more flexible (but, seems to be what most people expect with Git). Global ignore all files for _your_ editors and you'll never commit them into another repo again.

[the remark this answers is not in the document]
Comment (Dreamsorcerer) [reply .gitignore:92]: What I mean is that gitignore in a repo should be related to that repo. `.vscode` has nothing to do with the project, it's related to someone's computer, and therefore should be ignored on their computer (if VS code were smarter, it might add this to a global ignore when it was installing itself).

The gitignore files we have at work are typically about 2 lines long. They just contain something like `/dist` which is created by the build tool in our repo. More importantly, these changes have nothing to do with 'Feature/value checker', it can just be a little frustrating when people try to sneak in stylistic changes etc. to a project in a random PR.

File: aiocache/decorators.py (hunk 1 above)
Comment (Dreamsorcerer) [reply aiocache/decorators.py:65]: So, my understanding is that when you do `f = lambda: ...`, you are essentially trying to give a name to a lambda (`f` in this example). But, when printing it out etc. you won't see that name. That is an antipattern.

In this case, we are not trying to assign a name to the lambda, it is the default value which would not have a name associated with it.

File: aiocache/decorators.py (hunk 4 above)
Comment (VitalyPetrov) [reply aiocache/decorators.py:299]: This approach also require you to know all the possible keys.
The original idea was to perform value-based filtering for resulting dict. However I am also not sure this functionality is needed

File: .gitignore (hunk 5 above)
[the remark this answers is not in the document]
Comment (VitalyPetrov) [reply .gitignore]: Removed from `.gitignore`

File: aiocache/decorators.py [hunk 6]
```diff
@@ -39,6 +39,10 @@ class cached:
     :param key_builder: Callable that allows to build the function dynamically. It receives
         the function plus same args and kwargs passed to the function.
         This behavior is necessarily different than ``BaseCache.build_key()``
+    :param value_checker: Callable that allows to control whether the return
+        value to be cached or not. The callable receives the return value of the
+        wrapped function and should return the boolean indicating whether
+        to cache the result or not.
```
[the remark this answers is not in the document]
Comment (VitalyPetrov) [reply aiocache/decorators.py]: Agree, looks more clear.
Replaced

File: aiocache/decorators.py [hunk 7]
```diff
@@ -111,6 +117,13 @@ async def decorator(
 
         result = await f(*args, **kwargs)
 
+        if self.value_checker:
+            if not isinstance(self.value_checker(result), bool):
+                raise TypeError('value_checker callable should return bool')
+
+            if not self.value_checker(result):
```
[the remark this answers is not in the document]
Comment (VitalyPetrov) [reply aiocache/decorators.py]: Done

File: aiocache/decorators.py (hunk 2 above)
Comment (VitalyPetrov) [reply aiocache/decorators.py]: Done

File: aiocache/decorators.py (hunk 4 above)
Comment (VitalyPetrov) [reply aiocache/decorators.py]: The original idea was to provide an interface to perform value-based filtering (similar to your provided examples). So in both cases you need to know a range of all possible keys. However I m also not sure if this functionality is needed

**Comment by VitalyPetrov** [comment]
Updated the code, please check

File: aiocache/decorators.py (hunk 4 above)
Comment (Dreamsorcerer) [reply aiocache/decorators.py]: I don't see why you need the possible keys? In the above examples my cache function can look like:
```
def f(results):
    return (k for k in results.keys() if k.endswith("_nocache"))
```
or
```
def f(k, v):
    return k.endswith("_nocache")
```

Whereas the original does not allow filtering keys ending with `_nocache` unless you can name every single possible key beforehand.

Comment (Dreamsorcerer) [reply aiocache/decorators.py]: I'm feeling like the 2nd option is easier and clearer to understand, so I'm leaning towards that solution.

Comment (Dreamsorcerer) [reply aiocache/decorators.py]: Obviously, you can just as easily check values without caring about keys:
```
def f(k, v):
    return v is None
```

Comment (VitalyPetrov) [reply aiocache/decorators.py]: It seems I ve finally got your idea. Agree -- your variant is more generic and easy-to-use.

**Comment by Dreamsorcerer** [comment]
OK, looking good. If we can just get a couple of tests in (maybe in tests/acceptance/tests_decorator.py?), then we're good to merge.

**Comment by Dreamsorcerer** [comment]
Added cached_stampede to the tests to get full coverage, all good now.

Thanks a lot for that.
